BungeeCord-Patches/0060-Additional-DoS-mitigations.patch

```diff
@@ -240,6 +406,66 @@ index 738f0c92..ec33d337 100644
++import net.md_5.bungee.protocol.LimitedVarint21FrameDecoder;
+ import net.md_5.bungee.protocol.MinecraftDecoder;
+ import net.md_5.bungee.protocol.MinecraftEncoder;
+ import net.md_5.bungee.protocol.Protocol;
+@@ -82,7 +83,7 @@ public class PipelineUtils
+ 
+ 
+             try {
+-            BASE.initChannel( ch );
++            LIMITED_BASE.initChannel( ch ); // Waterfall - Use LIMITED_BASE (Additional DoS mitigations)
+             } catch (Exception e) {
+                 e.printStackTrace();
+                 ch.close();
+@@ -103,10 +104,12 @@ public class PipelineUtils
+             BungeeCord.getInstance().getPluginManager().callEvent(connectionInitEvent);
+         }
+     };
+-    public static final Base BASE = new Base();
++    public static final Base BASE = new Base(false); // Waterfall - Base class takes parameter now (Additional DoS mitigations)
++    public static final Base LIMITED_BASE = new Base(true); // Waterfall - Add LIMITED_BASE for connections form clients (Additional DoS mitigations)
+     private static final KickStringWriter legacyKicker = new KickStringWriter();
+     private static final Varint21LengthFieldPrepender framePrepender = new Varint21LengthFieldPrepender();
+     public static final String TIMEOUT_HANDLER = "timeout";
++    public static final String FRAMED_TIMEOUT_HANDLER = "framed-timeout"; // Waterfall - limit time we wait on complete (non-framed) packets
+     public static final String PACKET_DECODER = "packet-decoder";
+     public static final String PACKET_ENCODER = "packet-encoder";
+     public static final String BOSS_HANDLER = "inbound-boss";
+@@ -175,6 +178,13 @@ public class PipelineUtils
+ 
+     public static final class Base extends ChannelInitializer<Channel>
+     {
++        // Waterfall start - know if this a handler is for a client or a server (Additional DoS mitigations)
++        private final boolean fromClient;
++
++        public Base(boolean fromClient) {
++            this.fromClient = fromClient;
++        }
++        // Waterfall end
+ 
+         @Override
+         public void initChannel(Channel ch) throws Exception
+@@ -191,8 +201,9 @@ public class PipelineUtils
+             ch.config().setWriteBufferWaterMark( MARK );
+ 
+             ch.pipeline().addLast( TIMEOUT_HANDLER, new ReadTimeoutHandler( BungeeCord.getInstance().config.getTimeout(), TimeUnit.MILLISECONDS ) );
+-            ch.pipeline().addLast( FRAME_DECODER, new Varint21FrameDecoder() );
++            ch.pipeline().addLast( FRAME_DECODER, fromClient ? new LimitedVarint21FrameDecoder() : new Varint21FrameDecoder() ); // Waterfall - limit decoding length for client packets (Additional DoS mitigations)
+             ch.pipeline().addLast( FRAME_PREPENDER, framePrepender );
++            ch.pipeline().addLast( FRAMED_TIMEOUT_HANDLER, new ReadTimeoutHandler( BungeeCord.getInstance().config.getTimeout(), TimeUnit.MILLISECONDS ) ); // Waterfall - limit time we wait on complete (non-framed) packets
```

Comment:
This is outside the scope of this and potentially creates a race condition

Reply:
Ok, once the other points are cleared, I'll remove this one.

Reply:
Removed

Reply:
@Xernium Can you please elaborate why this would create a race condition?